### PR TITLE
feat(rpc): implement EthGetTransactionByBlockNumberAndIndex

### LINKED ISSUE #5020: EthGetTransactionByBlockNumberAndIndex
# Summary

This method is now supported in Lotus v1.31.0, and so we should support it as well.

# Completion Criteria

- [ ] Implement method
- [ ] Add api compare tests

# Additional Links & Resources

https://github.com/filecoin-project/lotus/pull/12618

### PR DESCRIPTION
**Pull request by virajbhartiya**
## Summary of changes

Changes introduced in this pull request:

implement EthGetTransactionByBlockNumberAndIndex

## Reference issue to close (if applicable)

Closes #5020

## Change checklist

- [X] I have performed a self-review of my own code,
- [X] I have made corresponding changes to the documentation. All new code adheres to the team's [documentation standards](https://github.com/ChainSafe/forest/wiki/Documentation-practices),
- [X] I have added tests that prove my fix is effective or that my feature works (if possible),
- [X] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/rpc/methods/eth.rs [hunk 1]
```diff
@@ -2015,19 +2015,51 @@ pub enum EthGetTransactionByBlockNumberAndIndex {}
...
-        (_p1, _p2): Self::Params,
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (block_number, tx_index): Self::Params,
     ) -> Result<Self::Ok, ServerError> {
-        // Lotus doesn't support this method (v1.29.0), so do we.
-        Err(ServerError::unsupported_method())
+        let height = block_number.0 as ChainEpoch;
+        let head = ctx.chain_store().heaviest_tipset();
+
+        if height > head.epoch() {
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:2032]: What about using `ensure` macro?

https://docs.rs/anyhow/latest/anyhow/macro.ensure.html

File: src/rpc/methods/eth.rs [hunk 2]
```diff
@@ -2015,19 +2015,51 @@ pub enum EthGetTransactionByBlockNumberAndIndex {}
...
+            return Err(anyhow::anyhow!("requested a future epoch (beyond \"latest\")").into());
+        }
+
+        let ts = ctx
+            .chain_index()
+            .tipset_by_height(height, head, ResolveNullTipset::TakeOlder)?;
+
+        let messages = ctx.chain_store().messages_for_tipset(&ts)?;
```
Comment (elmattic) [inline src/rpc/methods/eth.rs:2040]: Let's factor out the common code between the two methods?

**Comment by elmattic** [comment]
You have some CI error:

```bash
| Filecoin.EthGetTransactionByBlockNumberAndIndex (5) | Rejected("index 0 out of range: tipset contains 0 messages")                                  | Rejected("failed to get transaction at index 0: index 0 out of range: tipset contains 0 messages")
```

**Comment by elmattic** [comment]
Lotus method implementation is accepting values like "latest", "pending", etc? (see https://github.com/filecoin-project/lotus/pull/12618#issuecomment-2428101791)

What about us?

File: src/rpc/methods/eth.rs (hunk 1 above)
Comment (elmattic) [reply src/rpc/methods/eth.rs:2032]: Sorry, forget about this one. I don’t think using a `bail!` or `ensure!` macro is possible here.

File: src/rpc/methods/eth.rs (hunk 2 above)
Comment (elmattic) [reply src/rpc/methods/eth.rs:2040]: Ok, so it's not really possible.

**Comment by elmattic** [comment]
@virajbhartiya can you put the PR back to draft and only turn it to RFR when CI is ✅? Thanks!

In general, no green checkmark => no review.

**Comment by elmattic** [comment]
https://www.quicknode.com/docs/ethereum/eth_getTransactionByBlockNumberAndIndex

According to the specification, the method accepts only a string in hexadecimal format or tags for the block number. We can also support this by using a type like:

```
enum BlockNumberOrTag {
    BlockNumber(EthInt64),
    PredefinedBlock(Predefined),
}
```
Then we can create a conversion function to build a `BlockNumberOrHash` from this type and pass it to the `tipset_by_block_number_or_hash` function?

Of course, we can choose to diverge from the Ethereum specification, but this would complicate our efforts to align with a common OpenRPC standard.

See https://github.com/filecoin-project/FIPs/pull/1027.

File: CHANGELOG.md [hunk 3]
```diff
@@ -80,6 +80,11 @@ methods, fixes (notably to the garbage collection), and other improvements.
 - [#4701](https://github.com/ChainSafe/forest/issues/4701) Add support for the
   `Filecoin.EthGetTransactionByBlockHashAndIndex` RPC method.
 
+- [#5020](https://github.com/ChainSafe/forest/issues/5020) Add support for the
```
Comment (hanabi1224) [inline CHANGELOG.md:83]: I believe this needs to be moved to under `Forest unreleased` section

**Review by elmattic — APPROVED** [review]

**Review by hanabi1224 — APPROVED** [review]

**Pull request merged**